### DISCUSSION TITLE
Open Questions for a new ambitious project

### DISCUSSION THREAD

**Discussion by frosk1** [pkgxdev/pkgx#258] [Help] [answered]
Hey,

I just came across the tea project and the company behind it after reading an article about a new seed investment into the company.

First of all, I do like the ambitious project and the future plans. This is why I would like to ask the following questions:

Why using typescript instead of more low level richer tools like c, rust, c++?

Do you plan to include binaries in your remote repository / packages database / pantry?

Which build system do you use to build for example c++ projects in the pantry: cmake, bazel, etc.?

Will there be a recommandation or restriction which build system to use for tea projects?

Open source is the core feature and concept behind tea, how do you feel using typescript and github, both owned and maintained by microsoft for these planned ambitions (is it planned to move away from these technologies when going more decentralised using other ways of hosting the development of this software)?

**Comment by mfts** [comment, marked as answer]
Hey @frosk1, I can take some of those questions

> Do you plan to include binaries in your remote repository / packages database / pantry?

Our pantry repositories on GitHub will contain `package.yml`, which consists of **metadata and build instructions** for packages. Technically, you can download them and build the binaries locally if you'd like. However, we build the binaries in our CI/CD pipeline and upload them to our package registry. They are available with `tea`.

> Which build system do you use to build for example c++ projects in the pantry: cmake, bazel, etc.?

You specify the build instructions in the `package.yml` for your package. See my answer below. 

> Will there be a recommandation or restriction which build system to use for tea projects?

In [pantry.core](https://github.com/teaxyz/pantry.core), we have already packaged many languages and compilers. Furthermore, you can find build tools in [pantry.extra](https://github.com/teaxyz/pantry.extra). If you don't find what you need, please submit an issue or **build the required packages** yourself 😉

Have a look at existing `package.yml` files from projects such as [just.systems](https://github.com/teaxyz/pantry.extra/tree/main/projects/just.systems/package.yml) or [python.org](https://github.com/teaxyz/pantry.core/blob/main/projects/python.org/package.yml) to understand the breadth of build instructions. 

Finally, we have documented the structure of a `package.yml` file [here](https://github.com/teaxyz/pantry.zero/wiki/Structure-of-a-package.yml).

> Open source is the core feature and concept behind tea, how do you feel using typescript and github, both owned and maintained by microsoft for these planned ambitions (is it planned to move away from these technologies when going more decentralised using other ways of hosting the development of this software)?

GitHub at the moment is the place where most of open source projects live, so it was a natural decision for us to host our repository there as well. Git is already the most decentralized approach to software development, GitHub is just a way for projects to visualize/manage contributions more easily. That being said, we are always keeping an eye out for new tooling.

**Reply by frosk1** [reply]
Hey @mfts ,
thx for taking the time and answering some of the questions.


When asking about any favourites in build systems, I meant if you give a definition of what build system should be used
when writing new package.yml files. Assume I want to offer a c++ open source project within tea, should I use bazel or
cmake within the package.yml file, is there a recommandation or restriction and what is your default build system that you
use for building c++ projects in your core projects? (I saw you are using cmake most of the times)


With the last one I rather meant, if you deciding to switch to Gitlab for example, rather then Github.

I am a fan of github do not get me wrong here. I am just asking out of a pure open source idea in mind. Since
the goals of this company are ambitiouts,  putting open source developers first and get them paid etc.
the technology you are using should also be 100% open source funded, and managed rather then in the hands of
companies like Microsoft. And there are at least popular open core alternatives like "GitLab".

**Reply by frosk1** [reply]
Would love to hear reasons why choosing typescript for this kind of project? :)

**Reply by mfts** [reply]
> Why using typescript instead of more low level richer tools like c, rust, c++?

Rust is definitely great and we are fans of Rust internally. However, for a project like ours that needs speed of development in the beginning, the Typescript ecosystem is just the vastly bigger one and let's us iterate much faster as a team than c, c++ or even rust. 

We do see a future when tea/cli is more stable and we port tea/cli over to Rust. 
This will make the binary smaller and run slightly faster.

**Reply by mfts** [reply]
> Since the goals of this company are ambitiouts, putting open source developers first and get them paid etc. the technology you are using should also be 100% open source funded, and managed rather then in the hands of companies like Microsoft.

Our vision is certainly to remunerate open source projects and contributors equitably for their work. 
However, the status quo is that some open source projects are maintained by companies. 
As developers we shouldn't disregard the benefits of some open source software or tooling, just because they are maintained by companies. We need to be practical until we can achieve our vision.

**Reply by frosk1** [reply]
Not sure if I agree on the last one. I think one might need to differentiate between companies like Microsoft and smaller startup companies.
But thanks for being open and answering my questions.

Good luck to the project. I guess I definitely can see myself making some contribution if it goes more into the c++ or Rust direction ;)

**Reply by mfts** [reply]
Thanks 🙏 

> I guess I definitely can see myself making some contribution if it goes more into the c++ or Rust direction ;)

You can already contribute today by packaging crates from Rust ecosystem or packages you use in C++. All packages will be contributed to [pantry.extra](https://github.com/teaxyz/pantry.extra)

**Comment by mxcl** [comment]
> Why using typescript instead of more low level richer tools like c, rust, c++?

We can move fast with TypeScript and `deno`. `deno` also let’s us build our sources to a single portable binary for all platforms easily.

Once tea/cli is stable we will consider porting to rust mainly to reduce our end-binary size by 10×.
